File: sdks/go/cmd/elydora/plugins/claudecode_io.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
)

func claudeConfigDirectory() (string, error) {
	configured, exists := os.LookupEnv("CLAUDE_CONFIG_DIR")
	if !exists {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("resolve home directory: %w", err)
		}
		return filepath.Join(home, ".claude"), nil
	}
	resolved, err := filepath.Abs(configured)
	if err != nil {
		return "", fmt.Errorf("resolve CLAUDE_CONFIG_DIR at %s: %w", configured, err)
	}
	return resolved, nil
}

func claudeSettingsPath() (string, error) {
	directory, err := claudeConfigDirectory()
	if err != nil {
		return "", err
	}
	return filepath.Join(directory, claudeConfigFile), nil
}

func readClaudeDocument() (*claudeDocument, error) {
	filePath, err := claudeSettingsPath()
	if err != nil {
		return nil, err
	}
	if _, err := managedPhysicalDirectoryExists(
		filepath.Dir(filePath),
		"Claude Code configuration directory",
	); err != nil {
		return nil, err
	}
	snapshot, err := readManagedFile(
		filePath,
		"Claude Code user settings",
		maxManagedSourceBytes,
	)
	if err != nil {
		return nil, err
	}
	if snapshot == nil {
		return createClaudeDocument(filePath), nil
	}
	return parseClaudeDocument(filePath, snapshot.contents)
}

func prepareRenderedClaudeChange(
	rendered *claudeRenderedDocument,
) (*fileChange, error) {
	if rendered == nil || !rendered.changed {
		return nil, nil
	}
	return prepareSourceChange(
		rendered.document.filePath,
		"Claude Code user settings",
		rendered.document.raw,
		rendered.document.exists,
		rendered.next,
		0600,
		rendered.remove,
	)
}

func writeClaudeChanges(
	changes []*fileChange,
	label string,
	rename renameFunc,
	runtimeRoot string,
	agentDirectory string,
	settingsDirectory string,
) error {
	hasChanges := false
	for _, change := range changes {
		if change != nil {
			hasChanges = true
			break
		}
	}
	if !hasChanges {
		return nil
	}
	if agentDirectory != "" {
		if err := EnsurePrivateDirectory(runtimeRoot); err != nil {
			return err
		}
		if err := EnsurePrivateDirectory(agentDirectory); err != nil {
			return err
		}
	}
	if err := ensureManagedDirectory(
		settingsDirectory,
		"Claude Code configuration directory",
	); err != nil {
		return err
	}
	return writeChanges(changes, label, rename)
}

File: sdks/node/src/plugins/claudecode-io.ts
import os from 'node:os';
import path from 'node:path';
import {
  AGENT_KEY,
  CONFIG_FILE,
  createClaudeDocument,
  parseClaudeDocument,
  type ClaudeDocument,
  type ClaudeRuntimeContract,
  type RenderedClaudeDocument,
} from './claudecode-contract.js';
import { inspectPhysicalDirectory, readPhysicalFile } from './managed-files.js';
import { managedRuntimeFilesExist } from './managed-runtime-status.js';
import {
  commitManagedTransaction,
  prepareManagedFileChange,
} from './managed-transaction.js';

export function claudeConfigDirectory(): string {
  const configured = process.env.CLAUDE_CONFIG_DIR;
  return configured === undefined
    ? path.join(os.homedir(), '.claude')
    : path.resolve(configured);
}

export function claudeSettingsPath(): string {
  return path.join(claudeConfigDirectory(), CONFIG_FILE);
}

export async function readClaudeDocument(): Promise<ClaudeDocument> {
  const filePath = claudeSettingsPath();
  await inspectPhysicalDirectory(path.dirname(filePath), 'Claude Code configuration directory');
  const snapshot = await readPhysicalFile(filePath, 'Claude Code user settings');
  return snapshot
    ? parseClaudeDocument(filePath, snapshot.contents)
    : createClaudeDocument(filePath);
}

export async function writeClaudeDocument(rendered: RenderedClaudeDocument): Promise<void> {
  if (!rendered.changed) return;
  const change = await prepareManagedFileChange({
    filePath: rendered.document.filePath,
    label: 'Claude Code user settings',
    next: rendered.next,
    mode: 0o600,
    expectedSource: rendered.document.raw,
    verifyExpectedSource: true,
  });
  if (!change) return;
  await commitManagedTransaction({
    displayName: 'Claude Code',
    operation: 'uninstall',
    directories: [{
      path: path.dirname(rendered.document.filePath),
      label: 'Claude Code configuration directory',
    }],
    changes: [change],
  });
}

export async function claudeRuntimeFilesExist(
  contracts: readonly ClaudeRuntimeContract[],
): Promise<boolean> {
  for (const contract of contracts) {
    if (await managedRuntimeFilesExist(contract, AGENT_KEY)) return true;
  }
  return false;
}
